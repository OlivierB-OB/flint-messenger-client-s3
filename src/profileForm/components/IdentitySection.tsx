import FormControl from '@material-ui/core/FormControl';
import FormLabel from '@material-ui/core/FormLabel';
import FormGroup from '@material-ui/core/FormGroup';
import TextField from '@material-ui/core/TextField';
import React from 'react';
import { IFormField, IProfileFormFields } from '../types';

export interface IIdentitySectionProps {
  allowEmailEdition: boolean;
  email: IFormField<string>;
  firstName: IFormField<string>;
  lastName: IFormField<string>;
  update<T extends keyof IProfileFormFields>(field: T, value: IProfileFormFields[T]['value']): void;
}

export function IdentitySection({ allowEmailEdition, email, firstName, lastName, update }: IIdentitySectionProps) {
  return (
    <FormControl component="fieldset" fullWidth={true}>
      <FormLabel component="legend" style={{ margin: '1rem 0' }}>
        Your identity:
      </FormLabel>
      <FormGroup>
        <TextField
          label="Email"
          value={email.value}
          required={true}
          disabled={!allowEmailEdition}
          fullWidth={true}
          onChange={(event) => update('email', event.target.value)}
          {...(!email.isValid ? { error: true, helperText: email.error } : {})}
        />
        <TextField
          label="First name"
          value={firstName.value}
          required={true}
          fullWidth={true}
          onChange={(event) => update('firstName', event.target.value)}
          {...(!firstName.isValid ? { error: true, helperText: firstName.error } : {})}
        />
        <TextField
          label="Last name"
          value={lastName.value}
          required={true}
          fullWidth={true}
          onChange={(event) => update('lastName', event.target.value)}
          {...(!lastName.isValid ? { error: true, helperText: lastName.error } : {})}
        />
      </FormGroup>
    </FormControl>
  );
}
