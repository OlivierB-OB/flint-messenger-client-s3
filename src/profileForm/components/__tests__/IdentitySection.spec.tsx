import TextField from '@material-ui/core/TextField';
import React from 'react';
import { shallow } from 'enzyme';
import { IdentitySection } from '../IdentitySection';

describe('IdentitySection', () => {
  it('should display the email firstName and lastName value (email FYI only)', async () => {
    const data: any = {
      email: {
        value: 'email value',
        isValid: true,
      },
      firstName: {
        value: 'firstName value',
        isValid: true,
      },
      lastName: {
        value: 'lastName value',
        isValid: true,
      },
    };
    const component = shallow(<IdentitySection {...data} />);
    expect(component).toMatchSnapshot();
  });

  it('should display the email firstName and lastName value (email editable)', async () => {
    const data: any = {
      allowEmailEdition: true,
      email: {
        value: 'email value',
        isValid: true,
      },
      firstName: {
        value: 'firstName value',
        isValid: true,
      },
      lastName: {
        value: 'lastName value',
        isValid: true,
      },
    };
    const component = shallow(<IdentitySection {...data} />);
    expect(component).toMatchSnapshot();
  });

  it('should show error if any', async () => {
    const data: any = {
      allowEmailEdition: true,
      email: {
        value: 'email value',
        isValid: false,
        error: 'email error',
      },
      firstName: {
        value: 'firstName value',
        isValid: false,
        error: 'firstName error',
      },
      lastName: {
        value: 'lastName value',
        isValid: false,
        error: 'lastName error',
      },
    };
    const component = shallow(<IdentitySection {...data} />);
    expect(component).toMatchSnapshot();
  });

  it('should allow editing the email value', async () => {
    const data: any = {
      allowEmailEdition: true,
      email: {
        value: 'email value',
        isValid: true,
      },
      firstName: {
        value: 'firstName value',
        isValid: true,
      },
      lastName: {
        value: 'lastName value',
        isValid: true,
      },
      update: jest.fn(),
    };
    const component = shallow(<IdentitySection {...data} />);
    component
      .find(TextField)
      .at(0)
      .simulate('change', { target: { value: 'new value' } });
    expect(data.update).toHaveBeenCalledWith('email', 'new value');
  });

  it('should allow editing the firstName value', async () => {
    const data: any = {
      allowEmailEdition: true,
      email: {
        value: 'email value',
        isValid: true,
      },
      firstName: {
        value: 'firstName value',
        isValid: true,
      },
      lastName: {
        value: 'lastName value',
        isValid: true,
      },
      update: jest.fn(),
    };
    const component = shallow(<IdentitySection {...data} />);
    component
      .find(TextField)
      .at(1)
      .simulate('change', { target: { value: 'new value' } });
    expect(data.update).toHaveBeenCalledWith('firstName', 'new value');
  });

  it('should allow editing the lastName value', async () => {
    const data: any = {
      allowEmailEdition: true,
      email: {
        value: 'email value',
        isValid: true,
      },
      firstName: {
        value: 'firstName value',
        isValid: true,
      },
      lastName: {
        value: 'lastName value',
        isValid: true,
      },
      update: jest.fn(),
    };
    const component = shallow(<IdentitySection {...data} />);
    component
      .find(TextField)
      .at(2)
      .simulate('change', { target: { value: 'new value' } });
    expect(data.update).toHaveBeenCalledWith('lastName', 'new value');
  });
});
