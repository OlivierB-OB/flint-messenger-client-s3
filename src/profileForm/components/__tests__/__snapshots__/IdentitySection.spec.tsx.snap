// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdentitySection should display the email firstName and lastName value (email FYI only) 1`] = `
<WithStyles(ForwardRef(FormControl))
  component="fieldset"
  fullWidth={true}
>
  <WithStyles(ForwardRef(FormLabel))
    component="legend"
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    Your identity:
  </WithStyles(ForwardRef(FormLabel))>
  <WithStyles(ForwardRef(FormGroup))>
    <WithStyles(ForwardRef(TextField))
      disabled={true}
      fullWidth={true}
      label="Email"
      onChange={[Function]}
      required={true}
      value="email value"
    />
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="First name"
      onChange={[Function]}
      required={true}
      value="firstName value"
    />
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="Last name"
      onChange={[Function]}
      required={true}
      value="lastName value"
    />
  </WithStyles(ForwardRef(FormGroup))>
</WithStyles(ForwardRef(FormControl))>
`;

exports[`IdentitySection should display the email firstName and lastName value (email editable) 1`] = `
<WithStyles(ForwardRef(FormControl))
  component="fieldset"
  fullWidth={true}
>
  <WithStyles(ForwardRef(FormLabel))
    component="legend"
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    Your identity:
  </WithStyles(ForwardRef(FormLabel))>
  <WithStyles(ForwardRef(FormGroup))>
    <WithStyles(ForwardRef(TextField))
      disabled={false}
      fullWidth={true}
      label="Email"
      onChange={[Function]}
      required={true}
      value="email value"
    />
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="First name"
      onChange={[Function]}
      required={true}
      value="firstName value"
    />
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="Last name"
      onChange={[Function]}
      required={true}
      value="lastName value"
    />
  </WithStyles(ForwardRef(FormGroup))>
</WithStyles(ForwardRef(FormControl))>
`;

exports[`IdentitySection should show error if any 1`] = `
<WithStyles(ForwardRef(FormControl))
  component="fieldset"
  fullWidth={true}
>
  <WithStyles(ForwardRef(FormLabel))
    component="legend"
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    Your identity:
  </WithStyles(ForwardRef(FormLabel))>
  <WithStyles(ForwardRef(FormGroup))>
    <WithStyles(ForwardRef(TextField))
      disabled={false}
      error={true}
      fullWidth={true}
      helperText="email error"
      label="Email"
      onChange={[Function]}
      required={true}
      value="email value"
    />
    <WithStyles(ForwardRef(TextField))
      error={true}
      fullWidth={true}
      helperText="firstName error"
      label="First name"
      onChange={[Function]}
      required={true}
      value="firstName value"
    />
    <WithStyles(ForwardRef(TextField))
      error={true}
      fullWidth={true}
      helperText="lastName error"
      label="Last name"
      onChange={[Function]}
      required={true}
      value="lastName value"
    />
  </WithStyles(ForwardRef(FormGroup))>
</WithStyles(ForwardRef(FormControl))>
`;
